### PR TITLE
8331577: RISC-V: C2 CountLeadingZerosV

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review this patch adding CountLeadingZerosV and CountTrailingZerosV instrinsics?
Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8331577](https://bugs.openjdk.org/browse/JDK-8331577): RISC-V: C2 CountLeadingZerosV (**Sub-task** - P4)
 * [JDK-8331578](https://bugs.openjdk.org/browse/JDK-8331578): RISC-V: C2 CountTrailingZerosV (**Sub-task** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19153/head:pull/19153` \
`$ git checkout pull/19153`

Update a local copy of the PR: \
`$ git checkout pull/19153` \
`$ git pull https://git.openjdk.org/jdk.git pull/19153/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19153`

View PR using the GUI difftool: \
`$ git pr show -t 19153`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19153.diff">https://git.openjdk.org/jdk/pull/19153.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19153#issuecomment-2102228401)

### CODE REVIEW

**Comment by Hamlin-Li** [comment]
NOTE: the reason why let dst and src share one register (i.e. `(vReg dst_src, vRegMask_V0 v0)`) in masked version is that for inactive elements, we should keep the origin value, neither `mu` or `ma` will do it.

BTW, I will also re-visit all existing masked version instructions to make sure it works as expected.

**Comment by RealFYang** [comment]
> NOTE: the reason why let dst and src share one register (i.e. `(vReg dst_src, vRegMask_V0 v0)`) in masked version is that for inactive elements, we should keep the origin value, neither `mu` or `ma` will do it.

Interesting. Is it specified anywhere? 

> BTW, I will also re-visit all existing masked version instructions to make sure it works as expected. tracked by https://bugs.openjdk.org/browse/JDK-8331992

I think this was considered before. What about the recently added ones like ReverseBytesV, PopCountVI/L?

**Comment by Hamlin-Li** [comment]
> > NOTE: the reason why let dst and src share one register (i.e. `(vReg dst_src, vRegMask_V0 v0)`) in masked version is that for inactive elements, we should keep the origin value, neither `mu` or `ma` will do it.
> 
> Interesting. Is it specified anywhere?

For the Semantics of `mu` or `ma`, it's https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#343-vector-tail-agnostic-and-vector-mask-agnostic-vta-and-vma. Based on this, we can deduce that here is a hidden bug.

> 
> > BTW, I will also re-visit all existing masked version instructions to make sure it works as expected. tracked by https://bugs.openjdk.org/browse/JDK-8331992
> 
> I think this issue was considered before when we were adding support for vector api. What about the recently added ones like ReverseBytesV, PopCountVI/L?

Yeh, I'm testing with a fix including those 2 intrinsics.

**Comment by RealFYang** [comment]
> > > BTW, I will also re-visit all existing masked version instructions to make sure it works as expected. tracked by https://bugs.openjdk.org/browse/JDK-8331992

Sorry for not being accurate. In fact, I mean requirement at the Java level. Why should we should keep the origin value for inactive elements?

**Comment by Hamlin-Li** [comment]
> > > > BTW, I will also re-visit all existing masked version instructions to make sure it works as expected. tracked by https://bugs.openjdk.org/browse/JDK-8331992
> 
> Sorry for not being accurate. In fact, I mean requirement at the Java level. Why should we keep the origin value of inactive elements from the input vector? I didn't notice such a requirement before.

I'm not sure about other places, but in vector APi, if you do operations with a mask, then the untouched (inactive in riscv vector insts) elements should be unmodified, i.e. same as original values.

**Comment by RealFYang** [comment]
> > Sorry for not being accurate. In fact, I mean requirement at the Java level. Why should we keep the origin value of inactive elements from the input vector? I didn't notice such a requirement before.
> 
> I'm not sure about other places, but in vector APi, if you do operations with a mask, then the untouched (inactive in riscv vector insts) elements should be unmodified, i.e. same as original values.

It will be helpful if you could point to the specific code or examples.

**Comment by Hamlin-Li** [comment]
> > > Sorry for not being accurate. In fact, I mean requirement at the Java level. Why should we keep the origin value of inactive elements from the input vector? I didn't notice such a requirement before.
> > 
> > 
> > I'm not sure about other places, but in vector APi, if you do operations with a mask, then the untouched (inactive in riscv vector insts) elements should be unmodified, i.e. same as original values.
> 
> It will be helpful if you could point to the specific code or examples.

For the example usage, please check the test code, e.g. https://github.com/openjdk/jdk/blob/master/test/jdk/jdk/incubator/vector/Byte64VectorTests.java#L5458
For the courterpart of this intrinsic in arm, please check https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/aarch64_vector.ad#L6391
Hope these information are helpful.

**Comment by RealFYang** [comment]
> > > > Sorry for not being accurate. In fact, I mean requirement at the Java level. Why should we keep the origin value of inactive elements from the input vector? I didn't notice such a requirement before.
> > > 
> > > 
> > > I'm not sure about other places, but in vector APi, if you do operations with a mask, then the untouched (inactive in riscv vector insts) elements should be unmodified, i.e. same as original values.
> > 
> > 
> > It will be helpful if you could point to the specific code or examples.
> 
> For the example usage, please check the test code, e.g. https://github.com/openjdk/jdk/blob/master/test/jdk/jdk/incubator/vector/Byte64VectorTests.java#L5458 For the courterpart of this intrinsic in arm, please check https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/aarch64_vector.ad#L6391 Hope these information are helpful.

Yeah, I think you are right. This is also reflected in the vector api source code like [1] [2].

[1] https://github.com/openjdk/jdk/blob/master/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java#L184
[2] https://github.com/openjdk/jdk/blob/master/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java#L231

File: src/hotspot/cpu/riscv/riscv_v.ad [hunk 1]
```diff
@@ -3759,14 +3761,14 @@ instruct vsignum_reg(vReg dst, vReg zero, vReg one, vRegMask_V0 v0) %{
 
 // -------------------------------- Reverse Bytes Vector Operations ------------------------
 
-instruct vreverse_bytes_masked(vReg dst, vReg src, vRegMask_V0 v0) %{
-  match(Set dst (ReverseBytesV src v0));
-  format %{ "vreverse_bytes_masked $dst, $src, v0" %}
+instruct vreverse_bytes_masked(vReg dst_src, vRegMask_V0 v0) %{
+  match(Set dst_src (ReverseBytesV dst_src v0));
+  format %{ "vreverse_bytes_masked $dst_src, $dst_src, v0" %}
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv_v.ad:3766]: Nit: I think we can use something more accurate like `vrev8.v` as the opcode name in format. That will be consistent with the RVV spec. Also I suggest `v0.t` instead of `v0` or `$v0` as the mask (Might deserve a separate PR for cleaning up other existing instructs). Similar for other newly added instructs.

Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/riscv_v.ad:3766]: Sure, let me do it later, tracked by https://bugs.openjdk.org/browse/JDK-8332030.

**Review by RealFYang — APPROVED** [review]

**Comment by Hamlin-Li** [comment]
Thanks @RealFYang for your reviewing.

/integrate
